### PR TITLE
feat(typescript): move typings inline

### PR DESCRIPTION
**Pull request by Haroenv**
I ran prettier on the file, and also removed duplicates, but check to make sure it works as expected @samouss

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,768 @@
+// Type definitions for algoliasearch-helper 2.26
+// Project: https://community.algolia.com/algoliasearch-helper-js/
+// Definitions by: Gordon Burgett <https://github.com/gburgett>
+//                 Haroen Viaene <https://github.com/haroenv>
+//                 Samuel Vaillant <https://github.com/samouss>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 2.8
```
Comment (samouss) [inline index.d.ts]: We can remove those lines now it's part of the project.

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,768 @@
...
+
+import { Client } from 'algoliasearch';
+import { EventEmitter } from 'events';
+import algoliasearch from 'algoliasearch';
```
Comment (samouss) [inline index.d.ts]: We can group the imports together.

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,768 @@
...
+ * Helper. It is a also a function that instantiate the helper.
+ * To use the helper, you also need the Algolia JS client v3.
+ * @param client an AlgoliaSearch client
+ * @param index the name of the index to query
+ * @param opts
+ */
+declare function algoliasearchHelper(
+  client: Client,
+  index: string,
+  opts: algoliasearchHelper.QueryParameters
```
Comment (samouss) [inline index.d.ts]:
```suggestion
  opts?: algoliasearchHelper.QueryParameters
```

File: index.d.ts [hunk 4]
```diff
@@ -0,0 +1,768 @@
...
+ * The algoliasearchHelper module is the function that will let its
+ * contains everything needed to use the Algoliasearch
+ * Helper. It is a also a function that instantiate the helper.
+ * To use the helper, you also need the Algolia JS client v3.
+ * @param client an AlgoliaSearch client
+ * @param index the name of the index to query
+ * @param opts
+ */
+declare function algoliasearchHelper(
+  client: Client,
```
Comment (samouss) [inline index.d.ts]: It's not true, it should be a `searchClient` interface not a full client.

File: index.d.ts [hunk 5]
```diff
@@ -0,0 +1,768 @@
...
+     * function promiseHandler(res) {
+     *   // res contains
+     *   // {
+     *   //   content : SearchResults
+     *   //   state   : SearchParameters (the one used for this specific search)
+     *   // }
+     * }
+     */
+    searchOnce(
+      options: QueryParameters
```
Comment (samouss) [inline index.d.ts]:
```suggestion
      options: SearchParameters
```

File: index.d.ts [hunk 6]
```diff
@@ -0,0 +1,768 @@
...
+    ): Promise<{ content: SearchResults; state: SearchParameters }>;
+    searchOnce(
+      options: QueryParameters,
```
Comment (samouss) [inline index.d.ts]:
```suggestion
      options: SearchParameters,
```

File: index.d.ts [hunk 7]
```diff
@@ -0,0 +1,768 @@
...
+    setQueryParameter(...args: any[]): any;
+
+    /**
+     * Set the whole state (warning: will erase previous state)
+     * @param newState the whole new state
+     * @return
+     * @fires change
+     * @chainable
+     */
+    setState(newState: QueryParameters): this;
```
Comment (samouss) [inline index.d.ts]:
```suggestion
    setState(newState: SearchParameters): this;
```

File: index.d.ts [hunk 8]
```diff
@@ -0,0 +1,768 @@
...
+     * state
+     * @example
+     * // Get the complete state as stored in the helper
+     * helper.getState();
+     * @example
+     * // Get a part of the state with all the refinements on attributes and the query
+     * helper.getState(['query', 'attribute:category']);
+     */
+    getState(): SearchParameters;
+    getState(filters: string[]): QueryParameters;
```
Comment (samouss) [inline index.d.ts]: IIRC we dropped this function inside the latest version.

File: index.d.ts [hunk 9]
```diff
@@ -0,0 +1,768 @@
...
+    hasRefinements(...args: any[]): any;
+    isExcluded(...args: any[]): any;
+    isDisjunctiveRefined(...args: any[]): any;
+    hasTag(...args: any[]): any;
+    isTagRefined(...args: any[]): any;
+    getIndex(...args: any[]): any;
+    getCurrentPage(...args: any[]): any;
+    getPage(...args: any[]): any;
+    getTags(...args: any[]): any;
+    getQueryParameter(...args: any[]): any;
```
Comment (samouss) [inline index.d.ts]: IIRC we dropped this function inside the latest version.

File: index.d.ts [hunk 10]
```diff
@@ -0,0 +1,768 @@
...
+     * returned containing only the requested fields, otherwise return the unfiltered
+     * state
+     * @example
+     * // Get the complete state as stored in the helper
+     * helper.getState();
+     * @example
+     * // Get a part of the state with all the refinements on attributes and the query
+     * helper.getState(['query', 'attribute:category']);
+     */
+    getState(): SearchParameters;
```
Comment (samouss) [inline index.d.ts]: IIRC we dropped this function inside the latest version.

File: index.d.ts [hunk 11]
```diff
@@ -0,0 +1,768 @@
...
+    getState(filters: string[]): QueryParameters;
+
+    getStateAsQueryString(...args: any[]): any;
```
Comment (samouss) [inline index.d.ts]: IIRC we dropped this function inside the latest version.

File: index.d.ts [hunk 12]
```diff
@@ -0,0 +1,768 @@
...
+    setStateFromQueryString(...args: any[]): any;
```
Comment (samouss) [inline index.d.ts]: IIRC we dropped this function inside the latest version.

File: index.d.ts [hunk 13]
```diff
@@ -0,0 +1,768 @@
...
+    getTags(...args: any[]): any;
+    getQueryParameter(...args: any[]): any;
+    getRefinements(...args: any[]): any;
+    getNumericRefinement(...args: any[]): any;
+    getHierarchicalFacetBreadcrumb(...args: any[]): any;
+    containsRefinement(...args: any[]): any;
+    clearCache(...args: any[]): any;
+    setClient(...args: any[]): any;
+    getClient(...args: any[]): any;
+    derive(...args: any[]): any;
```
Comment (samouss) [inline index.d.ts]: We use the `derive` function inside InstantSearch now. It would be nice to have it typed.

File: index.d.ts [hunk 14]
```diff
@@ -0,0 +1,768 @@
...
+    detachDerivedHelper(...args: any[]): any;
```
Comment (samouss) [inline index.d.ts]: We use the `detachDerivedHelper ` function inside InstantSearch now. It would be nice to have it typed.

File: index.d.ts [hunk 15]
```diff
@@ -0,0 +1,768 @@
...
+    hasPendingRequests(...args: any[]): any;
+  }
+
+  namespace AlgoliaSearchHelper {
```
Comment (samouss) [inline index.d.ts]: We have two namespace that have almost the same name. We should merge them:
  - `algoliasearchHelper`
  - `AlgoliaSearchHelper`

File: index.d.ts [hunk 16]
```diff
@@ -0,0 +1,768 @@
...
+      isRefined: boolean;
+    }
+
+    /**
+     * Structure of the data resolved by the
+     * [`searchForFacetValues()`](reference.html#AlgoliaSearchHelper#searchForFacetValues)
+     * promise.
+     */
+    interface FacetSearchResult {
+      facetHits: FacetSearchHit;
```
Comment (samouss) [inline index.d.ts]:
```suggestion
      facetHits: FacetSearchHit[];
```

File: index.d.ts [hunk 17]
```diff
@@ -0,0 +1,768 @@
...
+      processingTimeMS: number;
+    }
+  }
+
+  export interface QueryParameters extends algoliasearch.QueryParameters {
```
Comment (samouss) [inline index.d.ts]: This interface should have a different name, right now it's confusing. When we refer to `QueryParameters` it's not clear which one it is: the one from the client or the Helper? The Helper manages its own state with a `SearchParameters` class. This class accepts a POJO that looks like class but without the methods. Inside this object every attributes are optional (but not inside `SearchParameters`).

File: index.d.ts [hunk 18]
```diff
@@ -0,0 +1,768 @@
...
+     * When querying algolia, the values stored in this attribute will
+     * be translated into the `facetFilters` attribute.
+     */
+    hierarchicalFacetsRefinements?: {
+      [facet: string]: SearchParameters.FacetList;
+    };
+  }
+
+  type ValueTypes<T> = T extends Array<infer U> ? U : never;
+  type ManagedParameters = {
```
Comment (samouss) [inline index.d.ts]: There is probably simpler ways to express this.

File: index.d.ts [hunk 19]
```diff
@@ -0,0 +1,768 @@
...
+      attributeName: string;
+      name: string;
+      numericValue: number;
+      operator: string;
+      count: number;
+      exhaustive: boolean;
+    }
+  }
+
+  export {};
```
Comment (samouss) [inline index.d.ts]: Is it useful?

File: index.d.ts [hunk 20]
```diff
@@ -0,0 +1,768 @@
...
+      value: number | number[]
+    ): SearchParameters;
+    addTagRefinement(tag: string): SearchParameters;
+    clearRefinements(
+      attribute?:
+        | string
+        | ((value: any, attribute: string, type: string) => void)
+    ): SearchParameters;
+    clearTags(): SearchParameters;
+    filter(filters: string[]): any;
```
Comment (samouss) [inline index.d.ts]: IIRC we dropped this function inside the latest version.

**Comment by yannickcr** [comment]
Great! But it seems very hard to maintain and prone to errors :cry:.

Any plan to convert the helper to TypeScript so we can have the type definitions next to the code?

(not dismissing this PR, just asking for future evolution :wink:)

**Comment by Haroenv** [comment]
Yes, this is planned to move to TypeScript in the future @yannickcr, but for now we just want to be able to make changes synchronous to the major version of the helper

File: index.d.ts [hunk 21]
```diff
@@ -0,0 +1,1197 @@
...
+    getNumericRefinement(...args: any[]): any;
+    getHierarchicalFacetBreadcrumb(...args: any[]): any;
+    containsRefinement(...args: any[]): any;
+    clearCache(...args: any[]): any;
+    setClient(client: SearchClient): this;
+    getClient(): SearchClient;
+    derive(
+      deriveFn: (oldParams: SearchParameters) => SearchParameters
+    ): AlgoliaSearchHelper;
+    detachDerivedHelper(derivedHelper: AlgoliaSearchHelper): void;
```
[the remark this answers is not in the document]
Comment (Haroenv) [reply index.d.ts]: e508244

File: index.d.ts [hunk 22]
```diff
@@ -0,0 +1,1197 @@
...
+    getRefinements(...args: any[]): any;
+    getNumericRefinement(...args: any[]): any;
+    getHierarchicalFacetBreadcrumb(...args: any[]): any;
+    containsRefinement(...args: any[]): any;
+    clearCache(...args: any[]): any;
+    setClient(client: SearchClient): this;
+    getClient(): SearchClient;
+    derive(
+      deriveFn: (oldParams: SearchParameters) => SearchParameters
+    ): AlgoliaSearchHelper;
```
[the remark this answers is not in the document]
Comment (Haroenv) [reply index.d.ts]: e508244

File: index.d.ts [hunk 23]
```diff
@@ -0,0 +1,1197 @@
...
+  index: string,
+  opts?: algoliasearchHelper.PlainSearchParameters
+): algoliasearchHelper.AlgoliaSearchHelper;
+
+declare namespace algoliasearchHelper {
+  export const version: string;
+
+  export class AlgoliaSearchHelper extends EventEmitter {
+    state: SearchParameters;
+    lastResults: SearchResults;
```
[the remark this answers is not in the document]
Comment (Haroenv) [reply index.d.ts]: f41e710

File: index.d.ts [hunk 24]
```diff
@@ -0,0 +1,1197 @@
...
+    derivedHelpers: AlgoliaSearchHelper[];
```
[the remark this answers is not in the document]
Comment (Haroenv) [reply index.d.ts]: 20a2fed

File: index.d.ts [hunk 25]
```diff
@@ -0,0 +1,1197 @@
...
+     * @fires error
+     * @chainable
+     */
+    search(): this;
+
+    /**
+     * Gets the search query parameters that would be sent to the Algolia Client
+     * for the hits
+     */
+    getQuery(): algoliasearchHelper.PlainSearchParameters;
```
[the remark this answers is not in the document]
Comment (Haroenv) [reply index.d.ts]: 6071dfa

File: index.d.ts [hunk 26]
```diff
@@ -0,0 +1,1197 @@
...
+     * function promiseHandler(res) {
+     *   // res contains
+     *   // {
+     *   //   content : SearchResults
+     *   //   state   : SearchParameters (the one used for this specific search)
+     *   // }
+     * }
+     */
+    searchOnce(
+      options: SearchParameters
```
[the remark this answers is not in the document]
Comment (Haroenv) [reply index.d.ts]: 221ebef

File: index.d.ts [hunk 27]
```diff
@@ -0,0 +1,1197 @@
...
+     * @param [maxFacetHits] the maximum number values returned. Should be > 0 and <= 100
+     * @param [userState] the set of custom parameters to use on top of the current state. Setting a property to `undefined` removes
+     * it in the generated query.
+     * @return the results of the search
+     */
+    searchForFacetValues(
+      facet: string,
+      query: string,
+      maxFacetHits: number,
+      userState: SearchParameters
```
[the remark this answers is not in the document]
Comment (Haroenv) [reply index.d.ts]: 221ebef

File: index.d.ts [hunk 28]
```diff
@@ -0,0 +1,1197 @@
...
+     *   //   content : SearchResults
+     *   //   state   : SearchParameters (the one used for this specific search)
+     *   // }
+     * }
+     */
+    searchOnce(
+      options: SearchParameters
+    ): Promise<{ content: SearchResults; state: SearchParameters }>;
+    searchOnce(
+      options: SearchParameters,
```
[the remark this answers is not in the document]
Comment (Haroenv) [reply index.d.ts]: 221ebef

File: index.d.ts [hunk 29]
```diff
@@ -0,0 +1,1197 @@
...
+declare function algoliasearchHelper(
+  client: SearchClient,
+  index: string,
+  opts?: algoliasearchHelper.PlainSearchParameters
+): algoliasearchHelper.AlgoliaSearchHelper;
+
+declare namespace algoliasearchHelper {
+  export const version: string;
+
+  export class AlgoliaSearchHelper extends EventEmitter {
```
[the remark this answers is not in the document]
Comment (Haroenv) [reply index.d.ts]: In InstantSearch we also used the name AlgoliaSearchHelper, but it's correct that we use it interchangeably as a single word or as two

[the remark this answers is not in the document]
Comment (Haroenv) [reply index.d.ts]: the real class is called `AlgoliaSearchHelper` so we should leave it as is

File: index.d.ts (hunk 25 above)
[the remark this answers is not in the document]
Comment (samouss) [reply index.d.ts]: It looks like the issue is not fixed.

**Pull request merged**